message/consumer.go

```diff
@@ -6,92 +6,94 @@ import (
 }
 
-// Consume consume incoming kafka messages delegating to the appropriate eventHandler or handling errors.
-func Consume(dimensionExtractedConsumer KafkaMessageConsumer, insertedProducer InsertedProducer, eventHandler EventHandler, exitChannel chan error) error {
+// ErrorEventHandler handler for dealing with any error while processing an inbound message.
+type ErrorEventHandler interface {
+	Handle(instanceID string, err error, data log.Data)
+}
+
+// Consume run a consumer to process incoming messages.
+func Consume(ctx context.Context, consumer KafkaConsumer, producer CompletedProducer, eventHandler EventHandler, errorEventHandler ErrorEventHandler) {
 	go func() {
 		for {
 			select {
-			case consumedMessage := <-dimensionExtractedConsumer.Incoming():
-				consumedData := consumedMessage.GetData()
-				var dimensionsExtractedEvent event.DimensionsExtractedEvent
-				if err := schema.DimensionsExtractedSchema.Unmarshal(consumedData, &dimensionsExtractedEvent); err != nil {
-					exitChannel <- err
-					return
-				}
-				logData := map[string]interface{}{eventKey: dimensionsExtractedEvent}
-				log.Debug(eventRecieved, logData)
-
-				if err := eventHandler.HandleEvent(dimensionsExtractedEvent); err != nil {
-					log.ErrorC(eventHandlerErr, err, logData)
-					exitChannel <- err
-					return
-				}
-
-				logData[logKeys.InstanceID] = dimensionsExtractedEvent.InstanceID
-				log.Debug(eventHandlerSuccess, logData)
-
-				insertedEvent := event.DimensionsInsertedEvent{
-					FileURL:    dimensionsExtractedEvent.FileURL,
-					InstanceID: dimensionsExtractedEvent.InstanceID,
-				}
-
-				if err := insertedProducer.DimensionInserted(insertedEvent); err != nil {
-					exitChannel <- err
-					return
-				}
+			case consumedMessage := <-consumer.Incoming():
+				processMessage(consumedMessage.GetData(), producer, eventHandler, errorEventHandler)
 				consumedMessage.Commit()
-
-			case consumerError := <-dimensionExtractedConsumer.Errors():
-				log.ErrorC(conumserErrMsg, consumerError, log.Data{logKeys.ErrorDetails: consumerError})
-				dimensionExtractedConsumer.Closer() <- true
-				insertedProducer.Closer() <- true
-				exitChannel <- consumerError
-				return
-			case producerError := <-insertedProducer.Errors():
-				log.ErrorC(producerErrMsg, producerError, nil)
-				dimensionExtractedConsumer.Closer() <- true
-				insertedProducer.Closer() <- true
-				exitChannel <- producerError
+			case <-ctx.Done():
+				log.Info(consumerStoppedMsg, nil)
 				return
 			}
 		}
 	}()
-	err := <-exitChannel
-	log.ErrorC(errorRecieved, err, nil)
-	return err
+}
+
+func processMessage(consumedData []byte, producer CompletedProducer, eventHandler EventHandler, errorEventHandler ErrorEventHandler) {
+	var newInstanceEvent event.NewInstanceEvent
+	if err := schema.NewInstanceSchema.Unmarshal(consumedData, &newInstanceEvent); err != nil {
+		log.ErrorC(unmarshallErrMsg, err, nil)
+		errorEventHandler.Handle("", err, nil)
```

Comment:
An interesting case - we will have to discuss what the error reporter should do with errors that cannot identify an instance to associate with.